---
title: "GSE118614 Smart-Seq2"
output: html_notebook
---

This is an analysis for the dataset [GSE118614](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE118614) made available by Clark BS et al. in their paper [_Single-cell RNA-Seq Analysis of Retinal Development Identifies NFI Factors as Regulating Mitotic Exit and Late-Born Cell Specification_](https://www.sciencedirect.com/science/article/abs/pii/S0896627319303459).
We observe a seemingly conspicuous batch effect in figure 1B and figure S1E in their original report. Thus, we try to make a correction.

```{r}
rm(list=ls())
gc()

library(Matrix)
library(ggplot2)
library(Rtsne)
library(matrixStats)
```

# Overview of the dataset
We first overview the dataset.

```{r}
barcode.path <- "../data/raw/GSE118614/GSE118614_Smart_cells.tsv.gz"
features.path <- "../data/raw/GSE118614/GSE118614_Smart_genes.tsv.gz"
matrix.path <- "../data/raw/GSE118614/GSE118614_Smart_aggregate.mtx.gz"
mat <- readMM(file = matrix.path)
feature.names = read.delim(features.path, 
                           header = TRUE,
                           stringsAsFactors = FALSE)
barcode.names = read.delim(barcode.path, 
                           header = TRUE,
                           stringsAsFactors = FALSE)
```

## Cell features

```{r}
barcode.names
```

There are 3 time points, each has 3 plates. The original reports concludes that each time point has mostly unique cell types. 

```{r}
table(barcode.names$Age)
table(barcode.names$Plate)
table(barcode.names$CellType)
table(barcode.names$Age, barcode.names$CellType)
table(barcode.names$Plate, barcode.names$CellType)
```

### tSNE plot using the given tSNE coordinates
It's also not a bad idea to reproduce it for a fair comparison in the future. But let's first look at the genes...

```{r}
ggplot(barcode.names) + geom_point(aes(x=tSNE1_pos, y=tSNE2_pos, color=Plate))
```

```{r}
ggplot(barcode.names) + geom_point(aes(x=tSNE1_pos, y=tSNE2_pos, color=CellType))
```

## Gene features

```{r}
feature.names
```

```{r}
hist(feature.names$num_cells_expressed)
```


```{r}
mat <- as.matrix(mat)
rownames(mat) <- barcode.names$Cell_Name
colnames(mat) <- rownames(feature.names)
```

tSNE-dimension reduction was performed on the top principal components learned from high variance genes in cells captured at individual time points. Mclust version 5.4 (Scrucca et al., 2016) was used to cluster cells in tSNE-space at which point cell type identity of clusters was assigned based on expression of known marker genes for either retinal or non-retinal tissue.

```{r}
th = 2

sum(feature.names$num_cells_expressed >= th)

#mat2 = t(t(mat) / colMeans(mat))

set.seed(0)
tsne_res <- Rtsne(log2(mat[, feature.names$num_cells_expressed > th] + 1), pca_scale=T)

tsne_plot <- data.frame(x = tsne_res$Y[, 1], y = tsne_res$Y[, 2], Plate = barcode.names$Plate)

ggplot(tsne_plot) + geom_point(aes(x=x, y=y, color=Plate))
```

```{r}
source("../R/varactor.R")
```

```{r}
obj <- Varactor$new(data = list(expr=t(mat)), 
                    labels = list(expr = list(plate = barcode.names$Plate, type=barcode.names$CellType, age=barcode.names$Age)), 
                    what = 'raw')
```

```{r}
#obj$normalize()
obj$normalized <- obj$raw
obj$combine(sd_threshold = .0, cv_threshold = 1.)
obj$reduce(reduce_dim = 40)
```

```{r}
set.seed(0)
obj$define_metric("primary", "euclidean")$measure("primary")$embed("primary", "tsne")
obj$plot("primary", "embedding", "type")
obj$plot("primary", "embedding", "plate")
obj$plot("primary", "embedding", "age")
```


```{r}
set.seed(0)
obj$define_metric("alternative", "davidson", strata = 'plate')$measure("alternative")$embed("alternative", "tsne")
obj$plot("alternative", "embedding", "type")
obj$plot("alternative", "embedding", "plate")
obj$plot("alternative", "embedding", "age")

obj$labels$type2 <- obj$labels$type
obj$labels$type2[obj$labels$type2 %in% c('Early RPCs', 'Intermediate RPCs', 'Late RPCs')] = 'RPCs'
obj$plot("alternative", "embedding", "type2")
```

```{r}
obj$define_metric("alternative", "davidson", strata = 'plate')$measure("alternative")$embed("alternative", "umap")
obj$plot("alternative", "embedding", "type")
obj$plot("alternative", "embedding", "plate")
obj$plot("alternative", "embedding", "age")

obj$labels$type2 <- obj$labels$type
obj$labels$type2[obj$labels$type2 %in% c('Early RPCs', 'Intermediate RPCs', 'Late RPCs')] = 'RPCs'
obj$plot("alternative", "embedding", "type2")
```

