---
title: "Varactor analysis of 10X sequenced and Seq-Well sequenced PBMC dataset"
output: html_notebook
---
 
In this notebook, we provide an example that Varactor decreases the difference of sequencing technologies, 10X and Seq-Well.

```{r}
rm(list=ls())
library(matrixStats)
library(Rtsne)
library(dplyr)
library(ggplot2)
library(RColorBrewer)
library(Seurat)
library(expm)  
```


# Viewing the Seq-Well dataset
We first view the Seq-Well [dataset](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM2486333), which is already annotated.

Ref. Gierahn T. Seq-Well: portable, low-cost RNA sequencing of single cells at high throughput. Nature Methods 2017(14).

## Load Seq-Well dataset

```{r}
pbmc_well = list()
pbmc_well$expr <- read.csv("../data/raw/pbmc_seq_well/GSM2486333_PBMC.txt", sep = ' ', header = T, row.names = 1)
```

## Explore
Apparently, the gene is in HUGO naming and the cells are named after their types.
```{r}
pbmc_well$expr <- as.matrix(pbmc_well$expr)
dim(pbmc_well$expr)
pbmc_well$expr[1:5, 1:5]
```

We extract the cell type information.
```{r}
pbmc_well$cell_type <- unlist(lapply(colnames(pbmc_well$expr), function(x) unlist(strsplit(x, split = '_'))[1]))
table(pbmc_well$cell_type)

pbmc_well$mask <- (pbmc_well$cell_type != "Removed") & (pbmc_well$cell_type != "DC")
sum(pbmc_well$mask)

pbmc_well$expr <- pbmc_well$expr[, pbmc_well$mask]
pbmc_well$cell_type <- pbmc_well$cell_type[pbmc_well$mask]
pbmc_well$mask <- NULL
```

We draw a t-SNE plot for the Seq-Well data only.
```{r}
pbmc_well$tsne <- Rtsne(t(pbmc_well$expr))
plot(pbmc_well$tsne$Y, col=factor(pbmc_well$cell_type))
```

# Viewing and annotating the Seq-Well dataset
The 10X dataset is not annotated, so we follow the procedure in the Seq-Well paper to filter the genes and annotate it.

```{r}
pbmc.data <- Read10X(data.dir = "../data/raw/pbmc3k_10x/filtered_gene_bc_matrices/hg19/")
# Initialize the Seurat object with the raw (non-normalized data).
pbmc <- CreateSeuratObject(counts = pbmc.data, 
                           project = "pbmc3k")

pbmc <- NormalizeData(pbmc, 
                      normalization.method = "LogNormalize", 
                      scale.factor = 10000, 
                      verbose = FALSE)

pbmc <- FindVariableFeatures(pbmc, 
                             selection.method = "mean.var.plot", 
                             nfeatures = 2000, 
                             mean.cutoff=c(0.5, Inf), 
                             dispersion.cutoff=c(0.5, Inf),
                             verbose=FALSE
                             )
length(VariableFeatures(pbmc))
```

```{r}
pbmc <- ScaleData(pbmc, features = rownames(pbmc))
pbmc <- RunPCA(pbmc, features = VariableFeatures(object = pbmc))
pbmc <- RunTSNE(pbmc)
```


```{r}
pbmc <- FindNeighbors(pbmc, dims = 1:13, k.param=50)
pbmc <- FindClusters(pbmc, resolution = 0.75)
DimPlot(pbmc, reduction = "tsne")
```
## Cell Markers
We use cell markers to identify the clusters.
Genes whose names are different from the corresponding CD markers: MS4A1 = CD20, ITGAX = CD11c, IL3RA = CD123, NCAM1 = CD56.

### T cell markers
```{r}
FeaturePlot(pbmc, 
            features = c("CD3E", "CD3D", "CD3G"),
            reduction = "tsne")
```

```{r}
FeaturePlot(pbmc,
            features = c("CD4"),
            reduction = "tsne")
```

```{r}
FeaturePlot(pbmc,
            features = c("CD8A", "CD8B"),
            reduction = "tsne")
```


### B cell markers
```{r}
FeaturePlot(pbmc,
            features = c("CD19", "MS4A1"),
            reduction = "tsne")
```

### Dendritic cell markers
```{r}
FeaturePlot(pbmc, 
            features = c("ITGAX", "IL3RA"),
            reduction = "tsne")
```

### NK cell markers
```{r}
FeaturePlot(pbmc,
            features = c("NCAM1"),
            reduction = "tsne")
```

### Monocyte markers
```{r}
FeaturePlot(pbmc, slot='counts',
            features = c("CD14", "CD33"),
            reduction = "tsne")
```

## Overexpressed genes in each cluster
We double check the overexpressed genes in each cluster.

### Cluster 0 - CD4?
```{r}
cluster0.markers <- FindMarkers(pbmc, ident.1 = 0, min.pct = 0.25, test.use="wilcox", only.pos = T, verbose=FALSE)
cluster0.markers
```



### Cluster 1 - Myeloid?
```{r}
cluster1.markers <- FindMarkers(pbmc, ident.1 = 1, min.pct = 0.25, test.use="wilcox", only.pos = T, verbose=FALSE)
cluster1.markers
```

### Cluster 2 - B cell?
```{r}
cluster2.markers <- FindMarkers(pbmc, ident.1 = 2, min.pct = 0.25, test.use="wilcox", only.pos = T, verbose=FALSE)
cluster2.markers
```

### Cluster 3 - CD8 T? 
```{r}
cluster3.markers <- FindMarkers(pbmc, ident.1 = 3, min.pct = 0.25, test.use="wilcox", only.pos = T, verbose=FALSE)
cluster3.markers
```

### Cluster 4 - Dendritic?

```{r}
cluster4.markers <- FindMarkers(pbmc, ident.1 = 4, min.pct = 0.25, test.use="wilcox", only.pos = T, verbose=FALSE)
cluster4.markers
```

### Cluster 5 - NK?
```{r}
cluster5.markers <- FindMarkers(pbmc, ident.1 = 5, min.pct = 0.25, test.use="wilcox", only.pos = T, verbose=FALSE)
cluster5.markers
```

```{r}
pbmc <- RenameIdents(pbmc, '0'='CD4', '1'='Myeloid', '2'='Bcell', '3'='CD8', '4'='DC', '5'='NK')
DimPlot(pbmc, reduction = "tsne")
```

```{r}
pbmc_10x = list()
pbmc_10x$expr <- as.matrix(pbmc@assays$RNA@data)
pbmc_10x$cell_type <- unname(Idents(pbmc))

pbmc_10x$cell_mask <- pbmc_10x$cell_type != 'DC'
pbmc_10x$expr <- pbmc_10x$expr[, pbmc_10x$cell_mask]
pbmc_10x$cell_type <- pbmc_10x$cell_type[pbmc_10x$cell_mask]
```

```{r}
#pbmc_10x$gene_mask <- rowSds(pbmc_10x$expr) > 0.0
#pbmc_10x$expr2 <- pbmc_10x$expr[pbmc_10x$gene_mask, ]

#pbmc_10x$gene_mask <- rowSds(pbmc_10x$expr2) / rowMeans(pbmc_10x$expr2) > 0.5
#pbmc_10x$expr2 <- pbmc_10x$expr2[pbmc_10x$gene_mask, ]

#pbmc_well$mask <- rowSds(pbmc_well$expr) > 0.0
#pbmc_well$expr2 <- pbmc_well$expr[pbmc_well$mask, ]

#pbmc_well$mask <- rowSds(pbmc_well$expr2) / rowMeans(pbmc_well$expr2) > 0.5
#pbmc_well$expr2 <- pbmc_well$expr2[pbmc_well$mask, ]


#gene_common <- intersect(rownames(pbmc_10x$expr2), rownames(pbmc_well$expr2))
#length(rownames(pbmc_10x$expr2))
#length(rownames(pbmc_well$expr2))
#length(gene_common)

#pbmc_10x$expr2 <- pbmc_10x$expr2[gene_common, ]
#pbmc_well$expr2 <- pbmc_well$expr2[gene_common, ]
```


```{r}
#int <- list()

#int$weight <- 1.
```


```{r}
#pbmc_10x$sd2 <- rowSds(pbmc_10x$expr2)
#pbmc_well$sd2 <- rowSds(pbmc_well$expr2)

#pbmc_10x$mean2 <- rowMeans(pbmc_10x$expr2)
#pbmc_well$mean2 <- rowMeans(pbmc_well$expr2)

#int$expr <- cbind((pbmc_10x$expr2 - pbmc_10x$mean2) / pbmc_10x$sd2,
#              (pbmc_well$expr2 - pbmc_well$mean2) / pbmc_well$sd2)
#int$expr <- int$expr * int$weight

#int$cell_type <- c(as.character(pbmc_10x$cell_type), pbmc_well$cell_type)
#int$tech <- c(rep('10x', length(pbmc_10x$cell_type)), rep('well', length(pbmc_well$cell_type)))
```

```{r}
#source("../R/varactor_class.R")
```

```{r}
#obj <- Varactor$new(data = int$expr, labels = list(tech=int$tech, type=int$cell_type), what = "normalized")
#obj$reduce(reduce_dim = 50)
#obj$reduced <- temp1
#obj$define_metric(name = "primary", type = "euclidean")
```

```{r}
#obj$measure("primary")
```

```{r}
#obj$embed(name = "primary", type = "tsne")
```

```{r}
#obj$plot_embedding("primary", "type")
```


```{r}
#temp1 <- obj$reduced
```

```{r}
#obj <- Varactor$new(data = int$expr, labels = list(tech=int$tech, type=int$cell_type), what = "normalized")$reduce()
```

```{r}
#obj$define_metric("primary", "euclidean")$measure("primary")$embed("primary", "tsne")$plot_embedding("primary", "type", pch=20)
```

```{r}
#obj$define_metric("alternative", "davidson", strata = "tech")$measure("alternative")$embed("alternative", "tsne")$plot_embedding("alternative", "type", pch=20)
```

# Jointly analysis using Varactor

## Create a Varactor object
```{r}
source("../R/varactor_class.R")
obj2 <- Varactor$new(data = list(x10x = pbmc_10x$expr, well = pbmc_well$expr), 
                     labels = list(x10x = list(type = as.character(pbmc_10x$cell_type)), 
                                   well = list(type = as.character(pbmc_well$cell_type))
                                  )
                    )
```

## Normalize, combine and perform dimensionality reduction
```{r}
obj2$normalize()
obj2$combine()
obj2$reduce()
```

```{r}
#temp_normalized -> obj2$normalized
#temp_reduced -> obj2$reduced
#temp_combined -> obj2$combined
```

## UMAP analysis

### Euclidean distance
Coloring: cell types
```{r}
obj2$define_metric("primary", "euclidean")$measure("primary")$embed("primary", "umap")$plot_embedding("primary", "type", pch=20)
```

Coloring: sequencing technologies
```{r}
obj2$plot_embedding("primary", "sample", pch=20)
```

Apparently, difference in technologies are shown heavily in the embedding, hindering our exploration to the biological meanings.


### Davidson distance
We use davidson distance to control the difference introduced by the techonology differences.

```{r}
obj2$define_metric("alternative", "davidson", strata = "sample")$measure("alternative")$embed("alternative", "umap")$plot_embedding("alternative", "type", pch=20)
```

```{r}
obj2$plot_embedding("alternative", "sample", pch=20)
```

The cell types are more tightly clustered and the difference between the technologies are decreased.

## t-SNE analysis
The same effects also appears in the t-SNE plots.

### Euclidean distance.

```{r}
obj2$define_metric("primary", "euclidean")$measure("primary")$embed("primary", "tsne")$plot_embedding("primary", "type", pch=20)
```

```{r}
obj2$plot_embedding("primary", "sample", pch=20)
```

### Davidson distance

```{r}
obj2$define_metric("alternative", "davidson", strata = "sample")$measure("alternative")$embed("alternative", "tsne")$plot_embedding("alternative", "type", pch=20)
```
```{r}
obj2$plot_embedding("alternative", "sample", pch=20)
```

