---
title: "GSE118614 10x"
output: html_notebook
---

This is an analysis for the dataset [GSE118614](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE118614) made available by Clark BS et al. in their paper [_Single-cell RNA-Seq Analysis of Retinal Development Identifies NFI Factors as Regulating Mitotic Exit and Late-Born Cell Specification_](https://www.sciencedirect.com/science/article/abs/pii/S0896627319303459).



```{r}
rm(list=ls())
gc()
library(Matrix)
library(Seurat)
```


```{r}
library(mclust)
```

# Overview of the dataset
We first overview the dataset.

```{r}
barcode.path <- "../data/GSE118614/barcodes.tsv"
features.path <- "../data/GSE118614/genes.tsv"
matrix.path <- "../data/GSE118614/matrix.mtx"
mat <- t(readMM(file = matrix.path))
feature.names = read.delim(features.path, 
                           header = TRUE,
                           stringsAsFactors = FALSE)
barcode.names = read.delim(barcode.path, 
                           header = TRUE,
                           stringsAsFactors = FALSE)
```

```{r}
dim(mat)
```

```{r}
colnames(mat) <- barcode.names$barcode
rownames(mat) <- feature.names$id
```

```{r}
obj <- CreateSeuratObject(mat, project="retina")
rm(mat)
gc()
```

```{r}
obj
```

```{r}
obj$age <- barcode.names$age
obj$sample <- barcode.names$sample
obj$celltype <- barcode.names$umap2_CellType
```

```{r}
table(obj$celltype)
```

```{r}
obj <- obj[, obj$celltype %in% c("Amacrine Cells", "Bipolar Cells", "Cones", "Early RPCs", "Horizontal Cells", 
                               "Late RPCs", "Muller Glia", "Neurogenic Cells", "Photoreceptor Precursors", 
                               "Photoreceptor Precursors", "Retinal Ganglion Cells", "Rods")]
```

```{r}
short_celltype <- obj$celltype
short_celltype[short_celltype == 'Amacrine Cells'] = "Am"
short_celltype[short_celltype == 'Bipolar Cells'] = "BC"
short_celltype[short_celltype == 'Cones'] = "Cone"
short_celltype[short_celltype == 'Early RPCs'] = "ERPC"
short_celltype[short_celltype == 'Horizontal Cells'] = "HC"
short_celltype[short_celltype == 'Late RPCs'] = "LRPC"
short_celltype[short_celltype == 'Muller Glia'] = "MG"
short_celltype[short_celltype == 'Neurogenic Cells'] = "Ng"
short_celltype[short_celltype == 'Photoreceptor Precursors'] = "Ph"
short_celltype[short_celltype == 'Retinal Ganglion Cells'] = "RGC"
short_celltype[short_celltype == 'Rods'] = "Rod"
obj$short_celltype <- short_celltype
```


```{r}
obj <- NormalizeData(obj, verbose=FALSE)
obj <- FindVariableFeatures(obj, selection.method = "vst", nfeatures = 2000, verbose=FALSE)

obj <- ScaleData(obj, verbose=FALSE)
obj <- RunPCA(obj, features = VariableFeatures(object = obj), verbose=FALSE)

ElbowPlot(obj)
```


```{r}
#source("../R/varactor_seurat2.R")
source("../R/bcd_seurat.R")

# ptm = proc.time()
feature.time.dict = c(E11 = -8,
                      E12_rep1 = -7,
                      E14_rep1 = -5,
                      E14_rep2 = -5,
                      E16 = -4,
                      E18_rep2 = -1,
                      E18_rep3 = -1,
                      P0 = 0,
                      P14 = 14,
                      P2_rep2 = 2,
                      P2_rep3 = 2,
                      P5 = 5,
                      P8_rep1 = 8,
                      P8_rep2 = 8)

for (reg in c(0.01, 0.05, 0.1, 0.5, 1., 5, 10, 50, 100)) {
  print(reg)
  obj <- RunALT(object = obj, feature.unwanted = "sample", dims.use = 1:10, reduction.use = "pca", 
                feature.time.dict = feature.time.dict, reduction.name = "alt", reduction.key = "ALT_", reg=reg)
  
  # proc.time() - ptm
  #alt_metric <- intCriteria(obj@reductions$alt@cell.embeddings, 
  #                          as.integer(as.factor(obj$short_celltype)), 
  #                          "Silhouette")
  #alt_metric 
  
  write.csv(obj@reductions$alt@cell.embeddings, paste0('../output-embeddings/gse118614_bcd_', reg, '.csv'))
}
```


```{r}

```

