---
title: "GSE112294 10x"
output: html_notebook
---

This is an analysis for the dataset [GSE118614](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE118614) made available by Clark BS et al. in their paper [_Single-cell RNA-Seq Analysis of Retinal Development Identifies NFI Factors as Regulating Mitotic Exit and Late-Born Cell Specification_](https://www.sciencedirect.com/science/article/abs/pii/S0896627319303459).



```{r}
rm(list=ls())
gc()
library(Matrix)
library(Seurat)
```

```{r}
genes <- read.csv("../data/GSE112294/WagnerScience2018_genes.csv", row.names = 1)
metadata <- read.csv("../data/GSE112294/WagnerScience2018_metadata.csv", row.names = 1)
data <- readMM("../data/GSE112294/WagnerScience2018.mtx")
data = t(data)
rownames(data) <- rownames(genes)
colnames(data) <- rownames(metadata)
obj <- CreateSeuratObject(data, meta.data = metadata)
```

```{r}
obj <- NormalizeData(obj, verbose=FALSE)
obj <- FindVariableFeatures(obj, selection.method = "vst", nfeatures = 2000, verbose=FALSE)

obj <- ScaleData(obj, verbose=FALSE)
obj <- RunPCA(obj, features = VariableFeatures(object = obj), verbose=FALSE)

ElbowPlot(obj)
```

```{r}
# obj <- FindNeighbors(obj, reduction = "pca", dims = 1:10, verbose=FALSE)
# obj <- FindClusters(obj, resolution = 0.5, verbose=FALSE)
```

```{r}
obj <- RunUMAP(obj, dims = 1:10, verbose=FALSE)
```

```{r}
table(obj$TimeID)
```


```{r}
source("../R/bcd_seurat.R")

ptm = proc.time()
feature.time.dict = c(DEW001 = 18 ,
DEW003 = 18 ,
DEW010 = 10 ,
DEW011 = 14 ,
DEW012 = 18 ,
DEW021 = 24 ,
DEW032 = 10 ,
DEW033 = 10 ,
DEW034 = 10 ,
DEW035 = 14 ,
DEW036 = 14 ,
DEW037 = 14 ,
DEW038 = 18 ,
DEW039 = 18 ,
DEW040 = 18 ,
DEW041 = 18 ,
DEW042 = 6 ,
DEW043 = 6 ,
DEW044 = 6 ,
DEW045 = 6 ,
DEW046 = 8 ,
DEW047 = 8 ,
DEW048 = 8 ,
DEW049 = 8 ,
DEW050 = 4 ,
DEW051 = 4 ,
DEW052 = 24 ,
DEW053 = 24 ,
DEW054 = 24 ,
DEW055 = 24 ,
DEW056 = 24 ,
DEW057 = 24 ,
DEW101 = 24 ,
DEW102 = 24 ,
DEW103 = 24 ,
DEW104 = 24 ,
DEW105 = 24 ,
DEW106 = 24 ,
DEW107 = 24 ,
DEW108 = 24 ,
DEW109 = 24 ,
DEW110 = 24 ,
DEW158 = 24 ,
DEW159 = 24 ,
DEW160 = 24 ,
DEW161 = 24 ,
DEW162 = 24 ,
DEW163 = 24 ,
DEW164 = 24 ,
DEW165 = 24 ,
DEW166 = 24 ,
DEW167 = 24 ,
DEW168 = 24 ,
DEW169 = 24)

obj <- RunALT(object = obj, feature.unwanted = "library_id", dims.use = 1:15, reduction.use = "pca", 
              feature.time.dict = feature.time.dict, reduction.name = "alt", reduction.key = "ALT_")

proc.time() - ptm
#alt_metric <- intCriteria(obj@reductions$alt@cell.embeddings, 
#                          as.integer(as.factor(obj$short_celltype)), 
#                          "Silhouette")
#alt_metric 
```
```{r}
obj <- readRDS("../output-embeddings/gse112294_lib_bcd.RDS")
```


```{r}
obj <- RunUMAP(obj, reduction = "alt", dims = 1:15, verbose=FALSE, reduction.name = "umap2")
```

```{r}
saveRDS(obj, "../output-embeddings/gse112294_lib_bcd.RDS")
```


```{r}
write.csv(obj@reductions$alt@cell.embeddings, '../output-embeddings/gse112294_lib_bcd.csv')
write.csv(obj@reductions$pca@cell.embeddings, '../output-embeddings/gse112294_lib_pca.csv')

write.csv(obj@reductions$umap@cell.embeddings, '../output-embeddings/gse112294_lib_bcd_umap.csv')
write.csv(obj@reductions$umap2@cell.embeddings, '../output-embeddings/gse112294_lib_pca_umap.csv')
# write.csv(obj@meta.data, '../output-embeddings/gse112294_label.csv')

```

